### PR TITLE
Add a mechanism to override the expected .good file

### PR DESCRIPTION
**Pull request by ronawho**
Add support for `GOODFILE` and `test.goodfile` files to the test system to enable specifying the expected output file (the `.good` file).

Historically, `.good` files only supported a fixed set of config prefixes like `test.comm-none.good` and `test.tasks-qthreads.good` and a few other `chplenv` specific settings. However, we have some tests with unique output for non-chplenv settings -- things like distribution tests have different comm counts for cyclic and block.

Previously, we used a prediff to figure out which dist specific `.good` file to copy and did something like `mv test.block.good test.good`. In some upcoming work, we'd like start_test to be able to tell us which expected `.good` file didn't match what actual output to either create a diff or provide some tooling to help accept the new output when you have a change that impacts the output for a lot of test. In order for that to work, the testing system needs to know the real on-disk files being diffed and that's what this new feature provides.

Part of Cray/chapel-private#1364

### CODE REVIEW

File: util/test/sub_test.py [hunk 1]
```diff
@@ -2311,6 +2322,24 @@ def filter_errors(output_in, pre_exec_output, execgoodfile, execlog):
                                              stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
                         sys.stdout.write(p.communicate()[0])
 
+                    if globalGoodfile:
```
Comment (stonea) [inline util/test/sub_test.py:2325]: Not something I think must be done for this PR but there's a lot of (mostly) duplicate code here when it comes to processing these scripts (goodfile, prediff, etc.)

It'd be nice to factor this out into a general "execute this script and get me the resulting stdout" function.

**Review by stonea — COMMENTED** [review]
Seems like a neat change. I have a couple of thoughts:

* `.good` file names can also be set using comments in `.compopts` and `.execopts`.  How does this work if a test both does that and has a .goodfile? Does one take precedence over the other, will it error out?

  * Just brainstorming: I'm guessing that if the user sets a good filename in `.compopts` or `.execopts` it gets assigned to `explicitexecgoodfile`, which you then overwrite after executing the `.goodfile` and capturing its standard out.  Maybe we should error out if `explicitexecgoodfile != None` at this point? Or maybe we could pass in `explicitexecgoodfile` to the good file script to let it modify it as it sees fit?

* Could you also update `doc/rst/developer/bestPractices/TestSystem.rst` to describe goodfiles and how they work?

**Comment by ronawho** [comment]
> * `.good` file names can also be set using comments in `.compopts` and `.execopts`.  How does this work if a test both does that and has a .goodfile? Does one take precedence over the other, will it error out?
>
>   * Just brainstorming: I'm guessing that if the user sets a good filename in `.compopts` or `.execopts` it gets assigned to `explicitexecgoodfile`, which you then overwrite after executing the `.goodfile` and capturing its standard out.  Maybe we should error out if `explicitexecgoodfile != None` at this point? Or maybe we could pass in `explicitexecgoodfile` to the good file script to let it modify it as it sees fit?

Hmm, these are good questions. I had not considered these cases and was only thinking about a test with just a .goodfile. In the current code `.goodfile` takes precedence over a .good file specified in an `.execopts` file, which takes precedence over a .good file specified in a `.compopts` file.

And beyond that I should also note that the `.goodfile` support here only works at execution time and for instance does not run for compilation only tests.

Let me take a step back and see if it's possible to update these tests to specify the .good file in an execopts file instead instead of adding this new functionality.

**Comment by ronawho** [comment]
Don't think I'm going to get back to this anytime soon, closing in the meantime.

**Pull request closed without merging**